lib/internal/test_runner/test.js

```diff
@@ -574,11 +569,17 @@ class Test extends AsyncResource {
         return;
       }
 
-      await after();
       await afterEach();
+      if (this.hooks.after.length > 0) {
```

Comment:
I am not sure I understand why this is needed. it the condition (`this.hooks.after.length > 0`) is inside `runOnce` - it should succeed even if the after hook is defined inside the test

Reply:
Because when `after` was originally defined it was at the start of the test, and hooks are only attached to the test as it progresses. Now this doesn't need to go within `runOnce` because it only runs at the end of the test, but also handles all `after` hooks that are attached throughout the test.

There's a weird interaction with the hooks here. The other three hooks are all run by subtests using the parent's hooks. [The `before` hook is actually put through `runOnce` within the `createHook` method](https://github.com/nodejs/node/blob/main/lib/internal/test_runner/test.js#L429-L431) so that it only runs for the first subtest.

The test cannot (the way it is currently written, at least) know how many subtests it has, so  a subtest cannot reliably know it is the last to run and therefore also run the `after` hook.

So, we don't need `runOnce` as that won't do the right job for an `after` hook as it does for the `before` hook, and we can't define the `after` method up front because at that point the test has no hooks attached.

---

OK I just wrote all of the above, which I stand by. However I just got your point that since the condition is within an anonymous function passed to `runOnce` that it will only evaluate later.

But, `runOnce` was not being used for it's purpose to only run the function once, so it was a misdirect and the function was succeeding based on the fact that it was running as a callback function, not because it was supposed to be run once. 

I think this should either stay like this. Or we can strip the hooks of the `runOnce` method and just set up `after` and `afterEach` as anonymous functions in order to avoid the repetition in the try and catch.

Does all of this make sense?

Reply:
the `runOnce` is there to allow running `after` both inside `try` and inside `catch` (so if `after` throws for example - we wont run it again).

>  think this should either stay like this. Or we can strip the hooks of the runOnce method and just set up after and afterEach as anonymous functions in order to avoid the repetition in the try and catch.

I dont think we should perform any changes with the `after` hook

Reply:
Oh, good point. I didn't think about the case where the `after` hook itself throws. I'll change it back.